commit 9eeb52ae712e72141c4c1d173048a606ba8f42f6
Author: Akinobu Mita <akinobu.mita@gmail.com>
Date:   Thu Aug 10 15:23:53 2017 -0700

    fault-inject: fix wrong should_fail() decision in task context
    
    Commit 1203c8e6fb0a ("fault-inject: simplify access check for fail-nth")
    unintentionally broke a conditional statement in should_fail().  Any
    faults are not injected in the task context by the change when the
    systematic fault injection is not used.
    
    This change restores to the previous correct behaviour.
    
    Link: http://lkml.kernel.org/r/1501633700-3488-1-git-send-email-akinobu.mita@gmail.com
    Fixes: 1203c8e6fb0a ("fault-inject: simplify access check for fail-nth")
    Signed-off-by: Akinobu Mita <akinobu.mita@gmail.com>
    Reported-by: Lu Fengqi <lufq.fnst@cn.fujitsu.com>
    Tested-by: Lu Fengqi <lufq.fnst@cn.fujitsu.com>
    Cc: Dmitry Vyukov <dvyukov@google.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

diff --git a/lib/fault-inject.c b/lib/fault-inject.c
index 7d315fdb9f13..cf7b129b0b2b 100644
--- a/lib/fault-inject.c
+++ b/lib/fault-inject.c
@@ -110,10 +110,12 @@ bool should_fail(struct fault_attr *attr, ssize_t size)
 	if (in_task()) {
 		unsigned int fail_nth = READ_ONCE(current->fail_nth);
 
-		if (fail_nth && !WRITE_ONCE(current->fail_nth, fail_nth - 1))
-			goto fail;
+		if (fail_nth) {
+			if (!WRITE_ONCE(current->fail_nth, fail_nth - 1))
+				goto fail;
 
-		return false;
+			return false;
+		}
 	}
 
 	/* No need to check any other properties if the probability is 0 */